Marlin/src/module/stepper.cpp

```diff
@@ -2355,30 +2357,24 @@ hal_timer_t Stepper::block_phase_isr() {
         #endif
```

Comment:
I think this is one of the two main changes. It is the off by one error you have mentioned @mh-dm. I think you are right (wish I'd spotted it myself!). The acceleration phase lasts one step too long because `step_events_completed` is incremented before these `if` statements. And the deceleration phase begins one step too late. Acceleration is capped at `current_block->nominal_rate` so we don't see a speed discontinuity there but deceleration is never long enough and so there are acceleration spikes at the end of each deceleration.

This change looks correct to me.

Reply:
As for the "off-by-one" error, I believe it's caused more by the use of "`<=`" where "`<`" would have been more appropriate. However the value of `step_events_completed` is accumulated only *after* `pulse_phase_isr` and it isn't simply incremented by one. So, during `block_phase_isr` the `step_events_completed` value should be considered to be the zero-based index of the step currently being examined, not some later step.

The rename of "until"->"before" and "after"->"start" is a reflection of the changes from "`<=`" and "`>`" to "`<`" and "`>=`", combined with the assumption that, during `block_phase_isr` the value of `step_events_completed` is the _**zero-based**_ index of a step that has not yet been processed here previously.